Stm32 clock ctrl common pll alternate clock sources + related fixes

str4t0m opened the pull request:
This PR allows select STM32 SOCs to configure the PLL_P, PLL_Q, PLL_R outputs as alternate clock source.
The PLL outputs are enabled during initialization in case the respective divider is defined in dts.

The functionality is verified by extending the stm32_common_devices test with a test case that configures an alternate clock source for ADC1. Test configurations for STMG0, STM32G4, and STM32WL were changed accordingly.

Additionally, some related testcase and driver fixes are included:
- tests: drivers: clock_control: stm32_wl fix external clock dts
- drivers/clock_control: fix stm32 common pll_q divider setup
  Closes #47101 
- tests/drivers/clock_control: stm32_common_devices: lptim check disabled

Tested only on G0, G4, L4, WB, and WL targets.

str4t0m committed:
tests: drivers: clock_control: stm32_wl fix external clock dts

The nucleo_wl55jc according to the datasheet does have a
NT2016SF-32M-END5875A 32MHz TCXO as HSE, therefore needs
enable the "hse-tcxo;" property to work, this was not the case
for the clock_configuration/stm32_common_devices test cases.

Additionally, remove the comment about about ST-Link clock,
because the source is the tcxo and not the ST-Link.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
drivers/clock_control: stm32 common move pllx defines to common header

All these series share the same defines, and while they are not used
by all socs of the common-driver, this is not exptectd to lead to
any conflict.
By moving the defines they can also be used in clock_stm32_ll_common.c

Additionally, the stm32g0 pll_div define was renamed to pllm
in order to match the other series.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
dts: clock: stm32 common: bindings for pllp, pllq, pllr clock sources

This commit introduces bindings that allow socs compatible with the
stm32 common driver to use the pll outputs pll_p, pll_q, pll_r,
to be used as a clock source for peripherals.
These are to be used instead of the, now removed, STM32_SRC_PLLCLK.
Applies to: STM32G0, STM32G4, STM32L4, STM32L5, STM32WB, and STM32WL.

STM32F2, and STM32F4 are not considered: very similar but only very limited
configuration possible. Only I2S(EXTclock,PLL_R), MC01, MCO2,
PLL_Q for 48MHz clock, and PLL_P as PLLCLK.

Doesn't apply to STM32L0, and STM32L1 which only have PLLMUL and PLLDIV.

Doesn't apply to STM32F0, STM32F1, STM32F3 only having a PLLMUL,
and no divider.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
drivers/clock_control: stm32 common pll src support(g0,g4,l4,l5,wb,wl)

This commit adds support to select pll outputs as peripheral clock
sources to the stm32 common driver.
With this commit they are only available on
STM32G0, STM32G4, STM32L4, STM32L5, STM32WB, and STM32WL.

Support for STM32F2, and STM32F4, which also have p,q,r dividers,
is not enabled in this commit.

Also, stm32_clock_control_get_subsys_rate is extended to return
the configured frequency in case they are enabled, otherwise 0.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
drivers/clock_control: fix stm32 common pll_q divider setup

In the stm32_ll_common driver the PLL_Q divider should not
be directly written to RCC->PLLCFGR->PLLQ, but should be
translated to the matching register value.
i.e. shifting the value to the correct position of the register
is not enough.

This commit makes sure the divider value taken from device tree, is
correctly translated ot the matching register register value by
converting it to a STM32CUBE LL definition LL_RCC_PLLQ_DIV_xx.

Typcial divider to register value mapping:
G0, WB, WL:
Dividers 2-8 mapped to register values 1-7(0 reserved).
G4, L4, L5:
Dividers 2,4,6,8 mapped to reg values 0,1,2,3.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
drivers/clock_control: stm32 common: enable PLL_P/Q and set PLL_P div

This commit configures the PLL_P divider for SOCs compatible to the
stm32_ll_common driver in case a value was defined via a dts property.

Additionally, in case the divider value is defined in the device
tree, the respective pll output is enabled during initialization
in set_up_plls().

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
drivers/clock_control: stm32 common allow pll also when it is not sysclk

The SOC specific implementations of the clock_stm32_ll_common driver
included the PLL specific functions only when PLL was selected as sysclock.

This commit changes the condition from "STM32_SYSCLK_SRC_PLL"
to "defined(STM32_PLL_ENABLED)".
As a result the pll could also be used as peripheral clock source
in case it is not the sysclock.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
tests/drivers/clock_control: stm32 add adc-pll_p overlays(g0,g4,wl)

For the STM32G0, STM32G4, and STM32WL enable the adc node in one
configuration, and select the PLL_P output as its clock source.
PLL_P divider is chosen to be 20 to make sure it's a unique frequency.
- g0, and g4 have pll as sysclk
- wl has hse as sysclock

The test configurations and the overlay-files are renamed accordingly.
All overlays that don't specify an alternative clock source still
make sure that the adc node is "okay" to be able to perform basic test.
The basic test only turns on and off the gate clock without checking the
frequency.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
tests/drivers/clock_control: stm32_common_devices: add adc alt. clk. src

This commit adds a test case that configures an alternative clock source
for an ADC peripheral.

In case no alt clock is available, only the gating clock is enabled
and disabled.
Unlike the i2c and lptim test, the actual gating clock frequency is
not checked, because for the adc there seems to be no uniform way
to retrieve the frequency via the hal.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
tests/drivers/clock_control: stm32_common_devices: lptim check disabled

The test checks if the peripheral gating clock was correctly disabled
after the test, but accidentally the I2C_CLK was checked instead of the
LPTIM_CLK.

This commit fixes this by using __HAL_RCC_LPTIM1_IS_CLK_ENABLED instead.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

erwango reviewed (changes requested):
Thanks for these fixes and improvements.
A couple of comments, otherwise LGTM

str4t0m committed:
dt-bindings/clock: remove STM32_SRC_CLOCK_MIN / MAX definitions

The STM32_SRC_CLOCK_MIN and STM32_SRC_CLOCK_MAX defines
are not really needed because non valid clock sources are already
filtered out by the precompiler.

Only STM32_SRC_CLOCK_MIN was used once in code and can be replaced.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
drivers/clock_control: stm32 g0, g4, l4, remove get_pllout_frequency()

After replacing STM32_SRC_PLLCLK by the STM32_SRC_PLL_x sources
this function is no longer needed and are therefore removed.

Also, those functions returned a wrong frequency.
They should have used get_pllsrc_frequency() instead of get_pll_source()
to get the input frequency.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m commented on include/zephyr/drivers/clock_control/stm32_clock_control.h:
```diff
@@ -114,6 +114,7 @@
 	DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(pll), st_stm32l4_pll_clock, okay) || \
 	DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(pll), st_stm32u5_pll_clock, okay) || \
 	DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(pll), st_stm32wb_pll_clock, okay) || \
+	DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(pll), st_stm32wl_pll_clock, okay) || \
```
My mistake, removed.

str4t0m commented on include/zephyr/dt-bindings/clock/stm32g4_clock.h:
```diff
@@ -30,8 +30,11 @@
+#define STM32_SRC_PLL_P		0x008
+#define STM32_SRC_PLL_Q		0x009
+#define STM32_SRC_PLL_R		0x00a
+#define STM32_SRC_PLLCLK	0x00b
```
Not really.
I thought about removing it, but wasn't sure so I kept it initially. PLLCLK is now removed on these series.
As a result, get_pllout_frequency() is no longer needed on those series, and therefore also removed.

Also added a commit to remove STM32_SRC_CLOCK_MIN/STM32_SRC_CLOCK_MAX.

erwango replied:
good, thanks

erwango reviewed (approved):
Great work!

carlescufi reviewed (approved):
Based on the header changes
